cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -410,6 +410,9 @@ func isCoreDNSVersionSupported(client clientset.Interface) (bool, error) {
 	}
 
 	for _, pod := range coreDNSPodList.Items {
+		if pod.Status.Phase == v1.PodPending {
+			continue
+		}
```

Comment:
this would return a valid version for an invalid digest in a Pod that is pending.

/cc @rajansandeep 

should this block until the Pods are running?

Reply:
we also have a panic related to this area of the code:
https://github.com/kubernetes/kubeadm/issues/2267

Reply:
> should this block until the Pods are running?

Yes, I am leaning towards that. 
It seems this issue is caused due to either the coredns pod being in a `Pending` state(kubernetes/kubeadm#2267) or it being in a `ContainerCreating ` state (#94286).

Skipping the check if the coredns pod is not running might result in a false validation. (especially if all coredns pods are in not running state)

Reply:
If there is no running pod, then select the image name as the validated object and output a warning log?

Reply:
for the time being we should apply a waiting function to get all the Pods as "Running".

the alternative is to relax the version validation and only use the image version/tag.

Reply:
It probably does make sense to have some wait if the pods are stuck in a `ContainerCreating` state. However, I am certain that this may not be the case with `PodPending`. If a pod is stuck in that state for more than a few seconds, then there's something elsewhere that needs tackling. In that case it's best to just return a proper error as we are clearly trying to perform this operation on a cluster that's in a bad state.

Reply:
we can have a 10seconds timeout with checks every 500ms?

@zouyee please adjust the PR description to "Fixes" this issue instead:
https://github.com/kubernetes/kubeadm/issues/2267

we should include a PollImmediate call (see example here):
https://github.com/kubernetes/kubernetes/blob/3a2e96efff7860ffb06eeba6d8603be53cb5125e/cmd/kubeadm/app/util/apiclient/idempotency.go#L64

that goes to the list of pods:
- checks if they running
- have a valid container
- store the image ID of the container in slice.

iterate over the retrieved imageIDs and set isValidVersion = false if one of them does not matches.

@zouyee do you want to work on this fix in this PR?

Reply:
@rajansandeep @rosti just to double check, our preflight has downloaded the new coredns image at this point, so we don't have to wait for the download?

Reply:
> @zouyee do you want to work on this fix in this PR?

yes